### DISCUSSION TITLE
Dropwizard 4.0.2 + Hibernate + ContainerResponseFilter with NameBinding

### DISCUSSION THREAD

**Discussion by irineuruiz** [dropwizard/dropwizard#8063] [General]
Hi,

We are using dropwizard 4.0.2 with hibernate and we have developed a ContainerResponseFilter to log access to certain resources, for this we want to use the ContainerResponseFilter's NameBinding functionality.
As we want to save this in the database, we also want to use hibernate and for that UnitOfWorkAwareProxyFactory.

However, to our surprise, when we use UnitOfWorkAwareProxyFactory to create the ContainerResponseFilter instance, the NameBinding functionality stops working and all resources end up triggering this filter, including some that are read only, which then causes an error in UnitOfWork.
Is it possible to fix this in future versions of Dropwizard?

**Comment by zUniQueX** [comment]
Hi @irineuruiz. A change of the behavior would highly affect the current execution semantics.

At the moment, the `UnitOfWorkApplicationListener` opens a `Session` for the Jersey execution lifecycle and wraps a transaction around the execution of the resource method. If we would commit the transaction after the execution of the response filters, an exception thrown in the execution of a response filter would abort the whole transaction. But if you annotate a resource method with `@UnitOfWork`, you'd expect the transaction to commit, if no exception was thrown in the resource method.

The problem you're experiencing with the `UnitOfWorkAwareProxyFactory` is caused from the creation of the proxy class. This newly generated subclass doesn't inherit the `@NameBinding` annotation because it's not marked `@Inherited`. Therefore your annotation is not found on the proxied subclass and the name binding doesn't apply correctly.

I think the best you can do is to create a transaction yourself. The `UnitOfWorkAspect` commits the transaction after the execution of the resource method but doesn't close the `Session`. You can query it with `ManagedSessionContext.currentSession()` and then manually open a new transaction.

**Comment by irineuruiz** [comment]
Thanks @zUniQueX 

In fact, we solve it here by doing exactly that, we are opening a transaction and committing it.
I just thought it would be good to warn you because we didn't know that UnitOfWorkAwareProxyFactory would "break" the NameBinding functionality and our developers took a while to understand what was happening.

**Comment by zUniQueX** [comment]
@irineuruiz That's a good hint. But I think the`UnitOfWorkAwareProxyFactory` isn't known so widely and other users will probably not deal with proxies themselves, if they can avoid it.

If you think this case has an impact to more users, documentation updates are greatly appreciated.
